codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java

```diff
@@ -554,33 +583,37 @@ final void weight(short weight) {
          */
         final void takeChild(DefaultStream child, boolean exclusive, List<ParentChangedEvent> events) {
             DefaultStream oldParent = child.parent();
-            events.add(new ParentChangedEvent(child, oldParent));
-            notifyParentChanging(child, this);
-            child.parent = this;
 
-            if (exclusive && !children.isEmpty()) {
-                // If it was requested that this child be the exclusive dependency of this node,
-                // move any previous children to the child node, becoming grand children of this node.
-                for (DefaultStream grandchild : removeAllChildren().values()) {
-                    child.takeChild(grandchild, false, events);
+            if (oldParent != this) {
+                events.add(new ParentChangedEvent(child, oldParent));
+                notifyParentChanging(child, this);
+                child.parent = this;
+                // We need the removal operation to happen first so the prioritizableForTree for the old parent to root
+                // path is updated with the correct child.prioritizableForTree() value.
+                if (oldParent != null && oldParent.children.remove(child.id()) != null) {
+                    oldParent.totalChildWeights -= child.weight();
+                    if (!child.isDescendantOf(oldParent)) {
+                        oldParent.decrementPrioritizableForTree(child.prioritizableForTree());
+                        if (oldParent.prioritizableForTree() == 0) {
+                            removeStream(oldParent);
```

Comment:
Yes there is concern not so much about notifying the listener but more about corrupting the tree links.  Priority tree operations are happening and we could be removing elements from the tree...I want to avoid keeping extra state and queueing these events if possible.  I'm not sure if this is entirely legitimate but I have been trying to justify it as "these nodes don't belong in the tree and are only their because of their dependents require them to be there.  If at any point they are no longer required through association they should immediately be removed.".

I went through a few scenarios (particularly when the new parent is a descendent of the current node) and didn't see any immediate issues with this approach...but would appreciate if you could come up with a case that would be problematic.